StringRep.hs

```diff
@@ -13,6 +14,76 @@ import qualified Miso.String as MS
+data TermH = WordH (EPM.Holey Token) | BinderH Token | LamH TermH TermH | AppH TermH [TermH] deriving (Show, Eq)
+
+holey :: String -> EPM.Holey MS.MisoString
+holey ""       = []
+holey ('_':xs) = Nothing : holey xs
+holey xs       = Just (MS.toMisoString i) : holey rest
+  where (i, rest) = span (/= '_') xs
+
+mixfixTable :: [[(EPM.Holey MS.MisoString, EPM.Associativity)]]
+mixfixTable = (map . map) (first holey)
+  [ [("_->_",          EPM.RightAssoc)]
+  , [("_,_",           EPM.NonAssoc)]
+  , [("if_then_else_", EPM.RightAssoc)]
+  , [("_|-_:_",        EPM.NonAssoc)]
+  , [("_+_",           EPM.LeftAssoc)]
+  , [("_*_",           EPM.LeftAssoc)]
+  ]
+
+grammar :: EP.Grammar r (EP.Prod r Token Token TermH)
+grammar = mdo
+  identB    <- EP.rule $ BinderH <$> EP.satisfy isLegalBinder
+  identW    <- EP.rule $ (WordH . pure . Just) <$> EP.satisfy isLegalConst
+  atom      <- EP.rule $ identW
+                      <|> EP.namedToken LParen *> program <* EP.namedToken RParen
+  normalApp <- EP.rule $ atom
+                      <|> AppH <$> atom <*> some atom
+  mixfixApp <- EPM.mixfixExpression table normalApp (AppH . WordH)
+  program    <- EP.rule $ mixfixApp 
+                      <|> LamH <$> identB <*> program 
+  return program
+  where
+    table = map (map $ first $ map $ fmap (EP.namedToken . Word)) mixfixTable
+    illegalIdents = [s | xs <- mixfixTable , (ys, _) <- xs , Just s <- ys] -- TODO: switch to hashtable, but required hashable version for ghcjs and unordered-containers are conflicted. 
+    isLegalBinder x | Binder b <- x = not $ elem b illegalIdents 
+                    | otherwise = False
+    isLegalConst x | Word w <- x = not $ elem w illegalIdents 
+                   | otherwise = False
+
+concatMixfix :: EPM.Holey Token -> MS.MisoString
+concatMixfix xs = MS.ms $ B.toLazyText $ goCat xs
+goCat :: EPM.Holey Token -> B.Builder 
+goCat []            = B.fromText ""
+goCat (Nothing:xs)  = B.fromText "_" <> goCat xs
+goCat ((Just (Word x)):xs) = (B.fromText $ MS.fromMisoString x) <> goCat xs
+
+desugarAppH :: [Name] -> TermH -> [TermH] -> Term
+desugarAppH ctx t [] = error "Invalid Application operends"
+desugarAppH ctx t [x] = Ap (postProc ctx t) $ postProc ctx x
+desugarAppH ctx t (x:xs) = Ap (postProc ctx t) $ desugarAppH ctx x xs
+
+-- TODO: Add underline to operators before deBrujin indexing. E.g. when "+" is in the ctx, we actually need to search for "_+_"
+deBrujing :: [Name] -> MS.MisoString -> Term
+deBrujing ctx x | Just v <- elemIndex x ctx = LocalVar v
+                | otherwise = Const x
+
+postProc :: [Name] -> TermH -> Term
+postProc ctx (LamH (BinderH (Binder b)) t) = Lam (M b) $ postProc (b:ctx) t
+postProc ctx t = postProc' ctx t
+postProc' ctx (AppH t ts) = desugarAppH ctx t ts
+postProc' ctx t = postProc'' ctx t
+postProc'' ctx (WordH holeyOp) | length holeyOp > 1 = deBrujing ctx $ concatMixfix holeyOp
```

Comment:
Calling length twice in the guards is a bit inefficient

Reply:
Resolved. But would you like to explain why? I thought this will be "automatically" lazy.

Reply:
GHC might run common subexpression elimination here, but it's not guaranteed to happen.